src/pip/_internal/network/auth.py

```diff
@@ -173,10 +173,13 @@ def _get_url_and_credentials(self, original_url):
         # Use any stored credentials that we have for this netloc
         username, password = self.passwords.get(netloc, (None, None))
 
-        if username is None and password is None:
-            # No stored credentials. Acquire new credentials without prompting
-            # the user. (e.g. from netrc, keyring, or the URL itself)
-            username, password = self._get_new_credentials(original_url)
+        # still grab if different creds for same domain
+        username_candidate, password_candidate = self._get_new_credentials(original_url)
+        if username_candidate is not None:
+            username = username_candidate
+            # Accept password only if username has been setted
+            if password_candidate is not None:
+                password = password_candidate
```

Comment:
Since `self._get_new_credentials()` is now always called, this should be *above* the `self.passwords` block. This makes the code easier to read.

Reply:
sure